pytorch_lightning/metrics/metric.py

```diff
@@ -1,262 +1,168 @@
-
-        """
+    def __init__(
+        self,
+        compute_on_step: bool = True,
+        ddp_sync_on_step: bool = False,
+        process_group: Optional[Any] = None,
+    ):
         super().__init__()
-        self.name = name
-        self._dtype = torch.get_default_dtype()
-        self._device = torch.device("cpu")
 
-        self.reduce_group = reduce_group
+        self.ddp_sync_on_step = ddp_sync_on_step
+        self.compute_on_step = compute_on_step
+        self.process_group = process_group
+        self._to_sync = True
 
-        # Buffer for holding aggregated state after each batch
-        self._step_vals = []
+        self.update = self._wrap_update(self.update)
+        self.compute = self._wrap_compute(self.compute)
+        self._computed = None
 
-        # Register hooks
-        self.register_forward_pre_hook(self.input_convert)
-        self.register_forward_hook(self.output_convert)
-        self.register_forward_hook(self.ddp_reduce)
-        self.register_forward_hook(self.compute)
+        # initialize state
+        self._reductions = {}
+        self._defaults = {}
 
-    @staticmethod
-    def input_convert(self, data: Any):
+    def add_state(self, name: str, default, dist_reduce_fx: Optional[Union[str, Callable]] = None):
         """
-        Implement how the inputs should be casted before calling forward
+        Adds metric state variable. Only used by subclasses.
 
         Args:
-            data: input to forward method
-
-        Returns:
-            casted data
-        """
-        return data
+            name: The name of the state variable. The variable will then be accessible at ``self.name``.
+            default: Default value of the state; can either be a tensor or an empty list. The state will be
+                reset to this value when ``self.reset()`` is called.
+            dist_reduce_fx (Optional): Function to reduce state accross mutliple GPUs. If value is ``"sum"``,
+                ``"mean"``, or ``"cat"``, we will use ``torch.sum``, ``torch.mean``, and ``torch.cat`` respectively,
+                each with argument ``dim=0``.
+        """
+        if not isinstance(default, torch.Tensor) or (isinstance(default, list) and len(default) != 0):
+            raise ValueError(
+                "state variable must be a tensor or any empty list (where you can append tensors)"
+            )
+
+        if dist_reduce_fx == "sum":
+            dist_reduce_fx = lambda x: torch.sum(x, dim=0)
+        elif dist_reduce_fx == "mean":
+            dist_reduce_fx = lambda x: torch.mean(x, dim=0)
+        elif dist_reduce_fx == "cat":
+            dist_reduce_fx = lambda x: torch.cat(x, dim=0)
+        elif dist_reduce_fx is not None and not isinstance(dist_reduce_fx, Callable):
+            raise ValueError(
+                "`dist_reduce_fx` must be callable or one of ['mean', 'sum', 'cat', None]"
+            )
+
+        setattr(self, name, default)
```

Comment:
What about using `self.register_buffer` here instead of `setattr`?

Reply:
Alright I added register buffer if the state is a tensor, else it just uses setattr, this will allow the correct device sync behavior